File: ProjectAshfall/ProjectAshfall/Models/Heroes/Hero.swift
// Hero.swift
// ProjectAshfall
//
// The playable hero unit — a named survivor with stats, skills, and gear.

import Foundation

// MARK: - Hero Classification

/// Tactical role a hero fills on the squad.
enum HeroRole: String, Codable, CaseIterable {
    case assault
    case heavy
    case recon
    case medic
    case engineer
    case sniper
    case demolitions

    /// Short description shown in the hero detail screen.
    var roleDescription: String {
        switch self {
        case .assault:      return "Front-line fighter with balanced offense"
        case .heavy:        return "High-durability tank that absorbs punishment"
        case .recon:        return "Fast scout with high evasion and crit"
        case .medic:        return "Healer and support specialist"
        case .engineer:     return "Builder and gadget specialist"
        case .sniper:       return "Long-range precision damage dealer"
        case .demolitions:  return "Area-of-effect damage specialist"
        }
    }
}

/// Rarity tier that gates base-stat budgets and skill counts.
enum HeroRarity: String, Codable, CaseIterable, Comparable {
    case common
    case rare
    case elite
    case legendary

    private var sortOrder: Int {
        switch self {
        case .common:    return 0
        case .rare:      return 1
        case .elite:     return 2
        case .legendary: return 3
        }
    }

    static func < (lhs: HeroRarity, rhs: HeroRarity) -> Bool {
        lhs.sortOrder < rhs.sortOrder
    }

    /// Stat multiplier applied to base stat budgets.
    var statMultiplier: Double {
        switch self {
        case .common:    return 1.0
        case .rare:      return 1.15
        case .elite:     return 1.35
        case .legendary: return 1.6
        }
    }

    /// Maximum number of skill slots available at this rarity.
    var maxSkillSlots: Int {
        switch self {
        case .common:    return 3
        case .rare:      return 4
        case .elite:     return 5
        case .legendary: return 6
        }
    }
}

// MARK: - Gear Slot

/// Named equipment slot on a hero.
enum GearSlot: String, Codable, CaseIterable {
    case weapon
    case armor
    case accessory
    case mod
}

/// A piece of equipment that occupies a gear slot and provides stat bonuses.
struct GearItem: Codable, Identifiable, Hashable {
    let id: UUID
    let name: String
    let slot: GearSlot
    let healthBonus: Double
    let attackBonus: Double
    let defenseBonus: Double
    let speedBonus: Double
    let critChanceBonus: Double
    let rarity: HeroRarity
    let iconName: String

    init(
        id: UUID = UUID(),
        name: String,
        slot: GearSlot,
        healthBonus: Double = 0,
        attackBonus: Double = 0,
        defenseBonus: Double = 0,
        speedBonus: Double = 0,
        critChanceBonus: Double = 0,
        rarity: HeroRarity = .common,
        iconName: String = "gear_default"
    ) {
        self.id = id
        self.name = name
        self.slot = slot
        self.healthBonus = healthBonus
        self.attackBonus = attackBonus
        self.defenseBonus = defenseBonus
        self.speedBonus = speedBonus
        self.critChanceBonus = critChanceBonus
        self.rarity = rarity
        self.iconName = iconName
    }
}

// MARK: - Passive Trait

/// An innate trait that grants a persistent bonus or special behaviour.
struct PassiveTrait: Codable, Identifiable, Hashable {
    let id: UUID
    let name: String
    let description: String
    let statModifier: StatModifier?

    init(
        id: UUID = UUID(),
        name: String,
        description: String,
        statModifier: StatModifier? = nil
    ) {
        self.id = id
        self.name = name
        self.description = description
        self.statModifier = statModifier
    }
}

/// Flat or percentage modifier to a specific stat.
struct StatModifier: Codable, Hashable {
    let stat: ModifiableStat
    let flatBonus: Double
    let percentBonus: Double

    init(stat: ModifiableStat, flatBonus: Double = 0, percentBonus: Double = 0) {
        self.stat = stat
        self.flatBonus = flatBonus
        self.percentBonus = percentBonus
    }
}

/// Stats that can be targeted by modifiers.
enum ModifiableStat: String, Codable, CaseIterable {
    case health, attack, defense, speed, critChance, morale
}

// MARK: - Hero

/// A single named survivor that the player can recruit, level, equip, and deploy.
final class Hero: Codable, Identifiable, ObservableObject {

    // MARK: Identity

    let id: UUID
    var name: String
    var role: HeroRole
    var rarity: HeroRarity
    var portraitName: String
    var backstory: String

    // MARK: Progression

    /// Current level, clamped to 1...60.
    var level: Int {
        didSet { level = min(max(level, 1), 60) }
    }

    /// Star rank representing ascension tier, clamped to 1...5.
    var rank: Int {
        didSet { rank = min(max(rank, 1), 5) }
    }

    var experience: Double

    // MARK: Base Stats

    var health: Double
    var attack: Double
    var defense: Double
    var speed: Double

    /// Critical hit chance (0.0 to 1.0).
    var critChance: Double

    /// Morale affects combat performance (0 to 100).
    var morale: Double {
        didSet { morale = min(max(morale, 0), 100) }
    }

    // MARK: Abilities & Equipment

    var skills: [Skill]

    /// Gear currently equipped, keyed by slot.
    var equippedGear: [GearSlot: GearItem]

    /// The hero's unique signature weapon name (unlocked at rank 3).
    var signatureWeapon: String?

    /// Innate traits that provide passive bonuses.
    var passiveTraits: [PassiveTrait]

    /// IDs of heroes this hero has bonds with.
    var bondPartnerIDs: [UUID]

    // MARK: State

    var isUnlocked: Bool

    /// Chapter/act that unlocks this hero in the campaign. Nil = available from start.
    var unlockChapter: String?

    // MARK: - Computed Properties

    /// Total health including gear and bond bonuses.
    var effectiveHealth: Double {
        let gearBonus = equippedGear.values.reduce(0.0) { $0 + $1.healthBonus }
        let traitBonus = traitModifier(for: .health)
        return (health + gearBonus) * (1.0 + traitBonus)
    }

    /// Total attack including gear and bond bonuses.
    var effectiveAttack: Double {
        let gearBonus = equippedGear.values.reduce(0.0) { $0 + $1.attackBonus }
        let traitBonus = traitModifier(for: .attack)
        return (attack + gearBonus) * (1.0 + traitBonus)
    }

    /// Total defense including gear and bond bonuses.
    var effectiveDefense: Double {
        let gearBonus = equippedGear.values.reduce(0.0) { $0 + $1.defenseBonus }
        let traitBonus = traitModifier(for: .defense)
        return (defense + gearBonus) * (1.0 + traitBonus)
    }

    /// Total speed including gear and bond bonuses.
    var effectiveSpeed: Double {
        let gearBonus = equippedGear.values.reduce(0.0) { $0 + $1.speedBonus }
        let traitBonus = traitModifier(for: .speed)
        return (speed + gearBonus) * (1.0 + traitBonus)
    }

    /// Total crit chance including gear bonuses, clamped to 0...1.
    var effectiveCritChance: Double {
        let gearBonus = equippedGear.values.reduce(0.0) { $0 + $1.critChanceBonus }
        return min(1.0, critChance + gearBonus)
    }

    /// Combined power rating for squad strength calculations.
    var powerRating: Double {
        effectiveHealth * 0.3 + effectiveAttack * 0.25 +
        effectiveDefense * 0.2 + effectiveSpeed * 0.15 +
        effectiveCritChance * 50.0 + Double(level) * 2.0
    }

    /// Experience required to reach the next level.
    var experienceToNextLevel: Double {
        100.0 * pow(1.18, Double(level))
    }

    /// Skills available at the current level.
    var unlockedSkills: [Skill] {
        skills.filter { $0.unlockLevel <= level }
    }

    var isAlive: Bool { health > 0 }

    // MARK: - Initializer

    init(
        id: UUID = UUID(),
        name: String,
        role: HeroRole,
        rarity: HeroRarity,
        level: Int = 1,
        rank: Int = 1,
        experience: Double = 0,
        health: Double,
        attack: Double,
        defense: Double,
        speed: Double,
        critChance: Double = 0.05,
        skills: [Skill] = [],
        equippedGear: [GearSlot: GearItem] = [:],
        signatureWeapon: String? = nil,
        passiveTraits: [PassiveTrait] = [],
        bondPartnerIDs: [UUID] = [],
        backstory: String = "",
        isUnlocked: Bool = false,
        morale: Double = 50,
        portraitName: String = "hero_default",
        unlockChapter: String? = nil
    ) {
        self.id = id
        self.name = name
        self.role = role
        self.rarity = rarity
        self.level = min(max(level, 1), 60)
        self.rank = min(max(rank, 1), 5)
        self.experience = experience
        self.health = health
        self.attack = attack
        self.defense = defense
        self.speed = speed
        self.critChance = critChance
        self.skills = skills
        self.equippedGear = equippedGear
        self.signatureWeapon = signatureWeapon
        self.passiveTraits = passiveTraits
        self.bondPartnerIDs = bondPartnerIDs
        self.backstory = backstory
        self.isUnlocked = isUnlocked
        self.morale = min(max(morale, 0), 100)
        self.portraitName = portraitName
        self.unlockChapter = unlockChapter
    }

    // MARK: - Methods

    /// Adds experience and triggers level-ups as needed. Returns the number of levels gained.
    @discardableResult
    func addExperience(_ amount: Double) -> Int {
        guard amount > 0 else { return 0 }
        experience += amount
        var levelsGained = 0
        while level < 60 && experience >= experienceToNextLevel {
            experience -= experienceToNextLevel
            levelUp()
            levelsGained += 1
        }
        return levelsGained
    }

    /// Advances the hero by one level, applying stat growth based on role and rarity.
    func levelUp() {
        guard level < 60 else { return }
        level += 1

        let growth = statGrowthPerLevel()
        health  += growth.health
        attack  += growth.attack
        defense += growth.defense
        speed   += growth.speed
    }

    /// Equips a gear item in its designated slot.
    /// Returns the previously equipped item if any.
    @discardableResult
    func equip(_ item: GearItem) -> GearItem? {
        let previous = equippedGear[item.slot]
        equippedGear[item.slot] = item
        return previous
    }

    /// Removes and returns the gear in the specified slot.
    @discardableResult
    func unequip(_ slot: GearSlot) -> GearItem? {
        let item = equippedGear[slot]
        equippedGear[slot] = nil
        return item
    }

    /// Adjusts morale by the given delta, clamped to 0...100.
    func adjustMorale(by delta: Double) {
        morale = min(100, max(0, morale + delta))
    }

    // MARK: - Helpers

    /// Aggregate percentage bonus from passive traits for a given stat.
    private func traitModifier(for stat: ModifiableStat) -> Double {
        passiveTraits.compactMap { $0.statModifier }
            .filter { $0.stat == stat }
            .reduce(0.0) { $0 + $1.percentBonus }
    }

    /// Per-level stat growth tuned by role.
    private func statGrowthPerLevel() -> (health: Double, attack: Double, defense: Double, speed: Double) {
        let m = rarity.statMultiplier
        switch role {
        case .assault:      return (12 * m,  4.0 * m, 2.5 * m, 1.5 * m)
        case .heavy:        return (18 * m,  2.5 * m, 4.0 * m, 0.8 * m)
        case .recon:        return (8  * m,  3.0 * m, 1.5 * m, 3.0 * m)
        case .medic:        return (10 * m,  2.0 * m, 2.0 * m, 2.0 * m)
        case .engineer:     return (10 * m,  2.5 * m, 3.0 * m, 1.5 * m)
        case .sniper:       return (8  * m,  5.0 * m, 1.5 * m, 1.5 * m)
        case .demolitions:  return (10 * m,  4.5 * m, 2.0 * m, 1.2 * m)
        }
    }

    // MARK: - Codable

    enum CodingKeys: String, CodingKey {
        case id, name, role, rarity, portraitName, backstory
        case level, rank, experience
        case health, attack, defense, speed, critChance, morale
        case skills, equippedGear, signatureWeapon, passiveTraits, bondPartnerIDs
        case isUnlocked, unlockChapter
    }

    convenience init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.init(
            id: try c.decode(UUID.self, forKey: .id),
            name: try c.decode(String.self, forKey: .name),
            role: try c.decode(HeroRole.self, forKey: .role),
            rarity: try c.decode(HeroRarity.self, forKey: .rarity),
            level: try c.decode(Int.self, forKey: .level),
            rank: try c.decode(Int.self, forKey: .rank),
            experience: try c.decode(Double.self, forKey: .experience),
            health: try c.decode(Double.self, forKey: .health),
            attack: try c.decode(Double.self, forKey: .attack),
            defense: try c.decode(Double.self, forKey: .defense),
            speed: try c.decode(Double.self, forKey: .speed),
            critChance: try c.decode(Double.self, forKey: .critChance),
            skills: try c.decode([Skill].self, forKey: .skills),
            equippedGear: try c.decode([GearSlot: GearItem].self, forKey: .equippedGear),
            signatureWeapon: try c.decodeIfPresent(String.self, forKey: .signatureWeapon),
            passiveTraits: try c.decode([PassiveTrait].self, forKey: .passiveTraits),
            bondPartnerIDs: try c.decode([UUID].self, forKey: .bondPartnerIDs),
            backstory: try c.decode(String.self, forKey: .backstory),
            isUnlocked: try c.decode(Bool.self, forKey: .isUnlocked),
            morale: try c.decode(Double.self, forKey: .morale),
            portraitName: try c.decode(String.self, forKey: .portraitName),
            unlockChapter: try c.decodeIfPresent(String.self, forKey: .unlockChapter)
        )
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(name, forKey: .name)
        try c.encode(role, forKey: .role)
        try c.encode(rarity, forKey: .rarity)
        try c.encode(portraitName, forKey: .portraitName)
        try c.encode(backstory, forKey: .backstory)
        try c.encode(level, forKey: .level)
        try c.encode(rank, forKey: .rank)
        try c.encode(experience, forKey: .experience)
        try c.encode(health, forKey: .health)
        try c.encode(attack, forKey: .attack)
        try c.encode(defense, forKey: .defense)
        try c.encode(speed, forKey: .speed)
        try c.encode(critChance, forKey: .critChance)
        try c.encode(morale, forKey: .morale)
        try c.encode(skills, forKey: .skills)
        try c.encode(equippedGear, forKey: .equippedGear)
        try c.encodeIfPresent(signatureWeapon, forKey: .signatureWeapon)
        try c.encode(passiveTraits, forKey: .passiveTraits)
        try c.encode(bondPartnerIDs, forKey: .bondPartnerIDs)
        try c.encode(isUnlocked, forKey: .isUnlocked)
        try c.encodeIfPresent(unlockChapter, forKey: .unlockChapter)
    }
}

// MARK: - Hashable / Equatable

extension Hero: Hashable {
    static func == (lhs: Hero, rhs: Hero) -> Bool { lhs.id == rhs.id }
    func hash(into hasher: inout Hasher) { hasher.combine(id) }
}

File: src/models/Hero.ts
/**
 * Hero.ts — Hero system for Project Ashfall.
 *
 * Defines roles, rarities, skills, gear, stats, and the core Hero class
 * with levelling, equipment, and serialization support.
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export enum HeroRole {
  Assault = 'Assault',
  Heavy = 'Heavy',
  Recon = 'Recon',
  Medic = 'Medic',
  Engineer = 'Engineer',
  Sniper = 'Sniper',
  Demolitions = 'Demolitions',
}

export enum HeroRarity {
  Common = 'Common',
  Rare = 'Rare',
  Elite = 'Elite',
  Legendary = 'Legendary',
}

// ---------------------------------------------------------------------------
// Skill
// ---------------------------------------------------------------------------

export type SkillType = 'active' | 'passive' | 'ultimate';
export type TargetType = 'single' | 'area' | 'self' | 'allAllies' | 'allEnemies';
export type EffectType =
  | 'damage'
  | 'heal'
  | 'buff'
  | 'debuff'
  | 'dot'
  | 'shield'
  | 'stun'
  | 'taunt'
  | 'stealth'
  | 'deploy'
  | 'morale'
  | 'reveal'
  | 'hack'
  | 'knockback'
  | 'bleed'
  | 'burn'
  | 'explosiveDamage'
  | 'critBoost'
  | 'speedBoost'
  | 'defenseBoost'
  | 'attackBoost';

export interface Skill {
  id: string;
  name: string;
  description: string;
  type: SkillType;
  targetType: TargetType;
  cooldown: number;
  currentCooldown: number;
  damage: number;
  healAmount: number;
  duration: number;
  effectType: EffectType;
  unlockLevel: number;
  iconKey: string;
}

export interface SkillJSON {
  id: string;
  name: string;
  description: string;
  type: SkillType;
  targetType: TargetType;
  cooldown: number;
  currentCooldown: number;
  damage: number;
  healAmount: number;
  duration: number;
  effectType: EffectType;
  unlockLevel: number;
  iconKey: string;
}

// ---------------------------------------------------------------------------
// Gear
// ---------------------------------------------------------------------------

export type GearSlot = 'weapon' | 'armor' | 'accessory' | 'mod';

export interface GearItem {
  id: string;
  name: string;
  slot: GearSlot;
  rarity: HeroRarity;
  statBonuses: Partial<HeroStats>;
  description: string;
}

export interface GearItemJSON {
  id: string;
  name: string;
  slot: GearSlot;
  rarity: string;
  statBonuses: Partial<HeroStats>;
  description: string;
}

// ---------------------------------------------------------------------------
// Stats
// ---------------------------------------------------------------------------

export interface HeroStats {
  health: number;
  maxHealth: number;
  attack: number;
  defense: number;
  speed: number;
  critChance: number;   // 0-1
  critMultiplier: number; // e.g. 1.5 = 150 %
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const MAX_LEVEL = 60;
const MAX_RANK = 5;

/** XP needed to advance FROM `level` to `level + 1`. */
export function xpForLevel(level: number): number {
  return 100 + level * level * 10;
}

/** Rarity-based stat multiplier applied when computing effective stats. */
function rarityMultiplier(rarity: HeroRarity): number {
  switch (rarity) {
    case HeroRarity.Common:    return 1.0;
    case HeroRarity.Rare:      return 1.1;
    case HeroRarity.Elite:     return 1.25;
    case HeroRarity.Legendary: return 1.4;
  }
}

// ---------------------------------------------------------------------------
// Serialization shape
// ---------------------------------------------------------------------------

export interface HeroJSON {
  id: string;
  name: string;
  role: string;
  rarity: string;
  level: number;
  rank: number;
  experience: number;
  baseStats: HeroStats;
  gear: Record<string, GearItemJSON>;
  skills: SkillJSON[];
  signatureWeapon: GearItemJSON | null;
  passiveTraits: string[];
  bondPartnerIds: string[];
  backstory: string;
  isUnlocked: boolean;
  morale: number;
}

// ---------------------------------------------------------------------------
// Hero class
// ---------------------------------------------------------------------------

export class Hero {
  public id: string;
  public name: string;
  public role: HeroRole;
  public rarity: HeroRarity;
  public level: number;
  public rank: number;
  public experience: number;
  public baseStats: HeroStats;
  public gear: Map<GearSlot, GearItem>;
  public skills: Skill[];
  public signatureWeapon: GearItem | null;
  public passiveTraits: string[];
  public bondPartnerIds: string[];
  public backstory: string;
  public isUnlocked: boolean;
  public morale: number;

  constructor(params: {
    id: string;
    name: string;
    role: HeroRole;
    rarity: HeroRarity;
    level?: number;
    rank?: number;
    experience?: number;
    baseStats: HeroStats;
    gear?: Map<GearSlot, GearItem>;
    skills?: Skill[];
    signatureWeapon?: GearItem | null;
    passiveTraits?: string[];
    bondPartnerIds?: string[];
    backstory?: string;
    isUnlocked?: boolean;
    morale?: number;
  }) {
    this.id = params.id;
    this.name = params.name;
    this.role = params.role;
    this.rarity = params.rarity;
    this.level = Math.max(1, Math.min(MAX_LEVEL, params.level ?? 1));
    this.rank = Math.max(1, Math.min(MAX_RANK, params.rank ?? 1));
    this.experience = params.experience ?? 0;
    this.baseStats = { ...params.baseStats };
    this.gear = params.gear ?? new Map<GearSlot, GearItem>();
    this.skills = params.skills ?? [];
    this.signatureWeapon = params.signatureWeapon ?? null;
    this.passiveTraits = params.passiveTraits ?? [];
    this.bondPartnerIds = params.bondPartnerIds ?? [];
    this.backstory = params.backstory ?? '';
    this.isUnlocked = params.isUnlocked ?? false;
    this.morale = Math.max(0, Math.min(100, params.morale ?? 50));
  }

  // -----------------------------------------------------------------------
  // Stats
  // -----------------------------------------------------------------------

  /**
   * Compute effective stats = base * level-scaling * rarity multiplier
   *   + all gear bonuses.
   *
   * Level scaling: each level past 1 gives +2 % compounding.
   */
  getEffectiveStats(): HeroStats {
    const levelScale = 1 + (this.level - 1) * 0.02;
    const rm = rarityMultiplier(this.rarity);

    const eff: HeroStats = {
      health:         Math.round(this.baseStats.health * levelScale * rm),
      maxHealth:      Math.round(this.baseStats.maxHealth * levelScale * rm),
      attack:         Math.round(this.baseStats.attack * levelScale * rm),
      defense:        Math.round(this.baseStats.defense * levelScale * rm),
      speed:          Math.round(this.baseStats.speed * levelScale * rm),
      critChance:     this.baseStats.critChance,
      critMultiplier: this.baseStats.critMultiplier,
    };

    // Accumulate gear bonuses.
    for (const item of this.gear.values()) {
      this.applyBonuses(eff, item.statBonuses);
    }

    // Signature weapon bonus (always active if present).
    if (this.signatureWeapon) {
      this.applyBonuses(eff, this.signatureWeapon.statBonuses);
    }

    // Ensure health does not exceed maxHealth.
    eff.health = Math.min(eff.health, eff.maxHealth);

    return eff;
  }

  private applyBonuses(stats: HeroStats, bonuses: Partial<HeroStats>): void {
    if (bonuses.health !== undefined)         stats.health         += bonuses.health;
    if (bonuses.maxHealth !== undefined)       stats.maxHealth      += bonuses.maxHealth;
    if (bonuses.attack !== undefined)          stats.attack         += bonuses.attack;
    if (bonuses.defense !== undefined)         stats.defense        += bonuses.defense;
    if (bonuses.speed !== undefined)           stats.speed          += bonuses.speed;
    if (bonuses.critChance !== undefined)      stats.critChance     += bonuses.critChance;
    if (bonuses.critMultiplier !== undefined)  stats.critMultiplier += bonuses.critMultiplier;
  }

  // -----------------------------------------------------------------------
  // Experience / levelling
  // -----------------------------------------------------------------------

  /** XP required to reach the next level from the current level. */
  xpToNextLevel(): number {
    if (this.level >= MAX_LEVEL) return Infinity;
    return xpForLevel(this.level);
  }

  /** Whether the hero has accumulated enough XP to level up. */
  canLevelUp(): boolean {
    return this.level < MAX_LEVEL && this.experience >= this.xpToNextLevel();
  }

  /** Grant experience and auto-level as many times as possible. */
  addExperience(amount: number): void {
    if (amount <= 0 || this.level >= MAX_LEVEL) return;
    this.experience += amount;
    while (this.canLevelUp()) {
      this.levelUp();
    }
  }

  /**
   * Advance one level, consuming the required XP.
   * Callers should check `canLevelUp()` first.
   */
  levelUp(): void {
    if (!this.canLevelUp()) return;
    this.experience -= this.xpToNextLevel();
    this.level = Math.min(this.level + 1, MAX_LEVEL);
  }

  // -----------------------------------------------------------------------
  // Gear
  // -----------------------------------------------------------------------

  /** Equip a gear item, replacing whatever currently occupies its slot. */
  equip(item: GearItem): GearItem | null {
    const previous = this.gear.get(item.slot) ?? null;
    this.gear.set(item.slot, item);
    return previous;
  }

  /** Remove and return the item in the given slot (or null). */
  unequip(slot: GearSlot): GearItem | null {
    const item = this.gear.get(slot) ?? null;
    this.gear.delete(slot);
    return item;
  }

  // -----------------------------------------------------------------------
  // Skills
  // -----------------------------------------------------------------------

  /** Return skills whose unlock level is <= the hero's current level. */
  getSkillsAtLevel(): Skill[] {
    return this.skills.filter(s => s.unlockLevel <= this.level);
  }

  // -----------------------------------------------------------------------
  // Serialization
  // -----------------------------------------------------------------------

  toJSON(): HeroJSON {
    const gear: Record<string, GearItemJSON> = {};
    for (const [slot, item] of this.gear) {
      gear[slot] = serializeGearItem(item);
    }

    return {
      id: this.id,
      name: this.name,
      role: this.role,
      rarity: this.rarity,
      level: this.level,
      rank: this.rank,
      experience: this.experience,
      baseStats: { ...this.baseStats },
      gear,
      skills: this.skills.map(serializeSkill),
      signatureWeapon: this.signatureWeapon
        ? serializeGearItem(this.signatureWeapon)
        : null,
      passiveTraits: [...this.passiveTraits],
      bondPartnerIds: [...this.bondPartnerIds],
      backstory: this.backstory,
      isUnlocked: this.isUnlocked,
      morale: this.morale,
    };
  }

  static fromJSON(data: HeroJSON): Hero {
    const gear = new Map<GearSlot, GearItem>();
    for (const [slot, itemData] of Object.entries(data.gear)) {
      gear.set(slot as GearSlot, deserializeGearItem(itemData));
    }

    return new Hero({
      id: data.id,
      name: data.name,
      role: data.role as HeroRole,
      rarity: data.rarity as HeroRarity,
      level: data.level,
      rank: data.rank,
      experience: data.experience,
      baseStats: { ...data.baseStats },
      gear,
      skills: data.skills.map(deserializeSkill),
      signatureWeapon: data.signatureWeapon
        ? deserializeGearItem(data.signatureWeapon)
        : null,
      passiveTraits: [...data.passiveTraits],
      bondPartnerIds: [...data.bondPartnerIds],
      backstory: data.backstory,
      isUnlocked: data.isUnlocked,
      morale: data.morale,
    });
  }
}

// ---------------------------------------------------------------------------
// Serialization helpers
// ---------------------------------------------------------------------------

function serializeSkill(s: Skill): SkillJSON {
  return { ...s };
}

function deserializeSkill(d: SkillJSON): Skill {
  return { ...d };
}

function serializeGearItem(g: GearItem): GearItemJSON {
  return {
    id: g.id,
    name: g.name,
    slot: g.slot,
    rarity: g.rarity,
    statBonuses: { ...g.statBonuses },
    description: g.description,
  };
}

function deserializeGearItem(d: GearItemJSON): GearItem {
  return {
    id: d.id,
    name: d.name,
    slot: d.slot,
    rarity: d.rarity as HeroRarity,
    statBonuses: { ...d.statBonuses },
    description: d.description,
  };
}
